+sw_tests/+unit_tests/unittest_spinw_addatom.m

```diff
@@ -0,0 +1,150 @@
+classdef unittest_spinw_addatom < sw_tests.unit_tests.unittest_super
+
+    properties
+        swobj = [];
+        abstol = 10*eps
+    end
+    properties (TestParameter)
+        property_error = {{'Z', 'spinw:addatom:WrongInput'}, ...
+            {'A', 'spinw:addatom:WrongInput'}, ...
+            {'biso', 'spinw:addatom:WrongInput'}, ...
+            {'ox', 'spinw:addatom:WrongInput'}, ...
+            {'S', 'spinw:sw_valid:SizeMismatch'}};
+    end
+
+    methods (TestMethodSetup)
+        function setup_spinw_model(testCase)
+            testCase.swobj = spinw(); % default init
+        end
+    end
+
+    methods (Test)
+        function test_no_input_calls_help(testCase)
+            % Tests that if call addmatrix with no input, it calls the help
+            help_function = sw_tests.utilities.mock_function('swhelp');
+            testCase.swobj.addatom();
+            testCase.assertEqual(help_function.n_calls, 1);
+            testCase.assertEqual(help_function.arguments, ...
+                {{'spinw.addatom'}});
+        end
+               
+        function test_add_multiple_atom_throws_mismatch_parameter_size(testCase, property_error)
+            [prop, error] = property_error{:};
+            testCase.verifyError(...
+                @() testCase.swobj.addatom('r', [0 0;0 0.5;0 0.5], prop, 2), ...
+                error)
+        end
+        
+        function test_add_atom_fails_invalid_position_size(testCase)
+            testCase.verifyError(...
+                @() testCase.swobj.addatom('r', [0, 0]), ...
+                'MATLAB:NonIntegerInput')  % need a better error for this
+        end
+        
+        function test_add_atom_fails_without_position(testCase)
+            testCase.verifyError(...
+                @() testCase.swobj.addatom('S', 1), ...
+                'sw_readparam:MissingParameter')
+        end
+        
+        function test_add_atom_fails_with_negative_spin(testCase)
+            testCase.verifyError(...
+                @() testCase.swobj.addatom('r', [0; 0; 0], 'S', -1), ...
+                'spinw:addatom:WrongInput')
+        end
+        
+        function test_add_single_default_atom_with_only_position(testCase)
```

Comment:
> Is there a reason ff, b and biso aren't checked here? Maybe we should have a function which checks all the fields of a unit_cell struct? To avoid being verbose maybe it could check against default values unless specific fields are input? This could also be useful for the single_ion, matrix structs, but it might be cumbersome in some cases so don't feel like you have to do it for everything though

Yes I don't know how I missed those off! This is a good idea, perhaps I will have a struct in the class that is the default atom and I can use that for tests to avoid duplication of the default values all the time.

Reply:
> I also noticed that in addatom.m the docstring mentions bx and bn, but the input bx doesn't appear to actually do anything (unless I missed it) and bn gets stored as b... maybe we shouldn't change the behaviour at this point but maybe we could update the docstring so it correctly describes the behaviour

I think there's an error in the docs that has caused some confusion, they say that `bn` is "Not implemented yet." but both `bn` and `b` change the neutron scattering length. As you have found, it is `bx` that is not implemented! I think actually I will just implement it and ensure there are tests for all three (`bn`, `bx`,`b`). It is not ideal that `b` and `bn` change the same field (note if you supply both `b` and `bn` then it ignores `b`) - removing one of `b` or `bn` is a breaking change so perhaps the best thing to do is throw a warning if `b` and `bn` are different if both provided?

Reply:
Actually the x-ray scattering length is a function of $|q|$ and the incident photon energy $E_0$ - see Eq. 4.1 in http://dx.doi.org/10.1088/0034-4885/69/1/R05
Though I suppose we would actually factor out the q-dependent form-factor and probably ignore the absorption, $f^{\prime\prime}(E_0)$ - which leaves us with a part of the cross-section proportional to $Z$ (which we expect) but also a part $f^\prime(E_0)$ which I don't think we are handling here (although we can apparently ignore it's q dependence). 

So it's not clear how the x-ray scattering length in `unit_cell.b(2,:)` should be interpreted. Also a quick grep of the code base shows that only the neutron cross-section evaluation in `structfact.m` uses `unit_cell.b(1,:)` - the second row isn't accessed in any file as far as I can see - is that the case @mducle ?

So in `addatom` I'm thinking I will move the `Not yet implemented` to refer to `bx` and throw two additional warnings:
1.  When `bn and `b` are both supplied saying b will be ignored
2. When `bx` is supplied saying it is ignored (or write whatever the user passes to `unit_cell.b(2,:)` - but I think that's more misleading)

The other option is to remove `bx` input (breaking change)...

Reply:
> > Is there a reason ff, b and biso aren't checked here? Maybe we should have a function which checks all the fields of a unit_cell struct? To avoid being verbose maybe it could check against default values unless specific fields are input? This could also be useful for the single_ion, matrix structs, but it might be cumbersome in some cases so don't feel like you have to do it for everything though
> 
> Yes I don't know how I missed those off! This is a good idea, perhaps I will have a struct in the class that is the default atom and I can use that for tests to avoid duplication of the default values all the time.

I have done this for `addatom` and similar for `addg` and `addaniso` (i.e. testing all fields in single_ion struct) - but have not done this for `addmatrix` as the `matrix` struct has only 3 fields, two of which (`label` and `mat`) change between tests and the other field `color` is randomly assigned by default.

Reply:
> Actually the x-ray scattering length is a function of |Q| and the incident photon energy E0 - see Eq. 4.1 in http://dx.doi.org/10.1088/0034-4885/69/1/R05 Though I suppose we would actually factor out the q-dependent form-factor and probably ignore the absorption,f′′(E0)|Q|$ dependence).
> 
> So it's not clear how the x-ray scattering length in `unit_cell.b(2,:)` should be interpreted. Also a quick grep of the code base shows that only the neutron cross-section evaluation in `structfact.m` uses `unit_cell.b(1,:)` - the second row isn't accessed in any file as far as I can see - is that the case @mducle ?

So I think that the `.b` are only used in Q-independent structure factor calculations, with the `.ff` (form-factor) field used for Q-dependent calculations instead. 

Now, as you found, the X-ray scattering cross-section _always_ has a Q-dependent form factor (because it involves a photon-electron interaction and electrons are spread out), similarly to the neutron-magnetic (neutron-electronic spin) interaction which also has a Q-dependent form factor. 

Therefore, I think that the `.b(2,:)` part should _never_ be used and there is no point in ever setting `.bx` - I guess Sandor added it for symmetric reason...

So, my suggestion would be to keep the synonym between `.b` and `.bn` and say that if both are specified `.b` is ignored, and to deprecate `.bx` and remove it in a future version (e.g. 4.1 or 5.0)... we could also deprecate `.bn` in favour of `.b` or vice versa too...

Reply:
This sounds good to me, deprecate `bx`, and deprecate `b` or `bn` (and test that they emit a warning if used!). Is there a standard Matlab `DeprecationWarning` equivalent?

Reply:
Just added `warning('spinw:addatom:DeprecationWarning', message)` - hope that's OK